### DISCUSSION TITLE
<py-script> code should have a better interaction with python modules

### DISCUSSION THREAD

**Discussion by antocuni** [pyscript/pyscript#1612] [Proposals]
I think that the current way of executing python code is fundamentally problematic, and I'd like to propose a way to fix it.

However, before I can describe the problem, I think it's a good idea to do a recap of how Python treats modules, because it's very relevant to our problem. If you already know how it works, you can skip `Part 0` and jump directly to `Part 1`.

This is very related to the old discussion that we had in https://github.com/pyscript/pyscript/pull/503
/cc @fpliger and @JeffersGlass 

---

## Part 0: Python modules recap

**NOTE**: here I'm talking about "normal" CPython on a desktop, not pyodide.

In Python, the global scope doesn't exist. The language uses some bad naming, such as the `global` keyword and the `globals()` function, but the reality is that 99.9% of all python code is always executed in the context of a module (the 0.1% is for `exec()` and `eval()`).

When you do `python foo.py`, python creates a special module called `__main__` in which it executes the content of `foo.py`. Basically, something like this:
```python
mod = types.ModuleType('__main__')
sys.modules['__main__'] = mod
src = open('foo.py').read()
exec(src, mod.__dict__)
```

This works fine for small script, but it creates tons of problems as soon as someone tries to `import foo`: in that case, python creates a *new* module called `foo` and *re-executes* the code again. This causes tons of confusion and it's a well known python pitfall, because now you have two copies of all the functions and module-level variables.

For example, try to predict what happens when you execute `python foo.py` with the following:
```python
import foo

N = 0
def inc():
    global N
    N += 1
    print(f'module {__name__}, N = {N}')

print('hello, I am the module', __name__)
inc()
foo.inc()
```

The pitfall is so common that for complex program, a good practice is to have a very small entry point which just does `import myprogram; myprogram.main()` and leave all the code inside the `myprogram` package.

---

## Part 1: PyScript current behavior

PyScript relies a lot on pyodide's `runPython()` methods. Essentially, `runPython()` executes the given source code inside the `__main__` module, and `pyodide.globals` points to `sys.modules['__main__'].__dict__`.

We do the following:
1. we automatically add certain names (such as `display()`, `HTML()` and `pyscript`) to `pyodide.globals`
2. we execute all the inline code inside `<py-script>` tags using `runPython()`: this is why they have access to e.g. `display()`
3. for `<py-script src=...>`, we fetch the source and execute it again using `runPython()`. This is why you have access to `display()` & co.

This leads to all sort of confusing behavior, as exemplified next

### Problem 1: overwrite functions by mistake

The following page both buttons prints `BBB`, because the first `onclick` is overwritten by the second.

```html
<py-script>
    def onclick():
        print('AAA')
</py-script>

<button id="aaa" py-click="onclick()">AAA</button>
<button id="bbb">BBB</button>

<py-script src="main.py"></py-script>
```
```python
# main.py
@when('click', selector='#bbb')
def onclick():
    print('BBB')
```

### Problem 2: re-imported modules

This is basically the same pitfall which was shown in part 0, but with an unique pyscript twist:

```html
<py-script src="main.py"></py-script>
<py-script src="other.py></py-script>
```
```python
# main.py
class Character:
    def __init__(self):
        self.position = 0

    def move_forward():
        self.position += 1

PLAYER = Character()

@when('click', selector="#btn-right")
def onclick():
    PLAYER.move_forward()
```
```python
# other.py
import main
def do_something():
    # this is NOT what you think:
    # we now have main.PLAYER and __main__.PLAYER 😱
    main.PLAYER.move_forward()
```

### Problem 3: missing globals

This is a problem which was actually reported by a pyscript user:

```html
<py-script src="main.py></py-script>
```
```python
# main.py
import mylib

def foo():
    # this works, because we injected display into __main__.__dict__
    display('hello')
    mylib.bar()
```
```python
# mylib.py
def bar():
    # this doesn't work because display() is not available here
    display('world')
```

This is arguably very very confusing.
We **could** fix this by injecting `display`&co. into the `builtins` module. I don't think it's a good idea.

---

## Part 2: Proposal for a new behavior

I think that the only way to have a behavior which is sane, unsurprising and that plays well with the "general python behavior" is keep the equation "1 file <=> 1 module". I propose the following semantics:

1. we stop automagically injecting names into `__main__.__dict__`
2. inline `<py-script>` tags are executed into `__main__`, exactly as now
3. `<py-script src="foo.py">` becomes de-facto equivalent to:
```html
<py-script>
    fetch_and_save("foo.py")
    import foo
</py-script>
```

This semantics solves all the surprising behavior described above. It *slightly* reduces the ergonomics of some easy tasks, but let me try to address them below.

### Easy access to display() & co.

To call `display()` & co., you need to explicitly import them from the `pyscript` package. I think this can be easily solved by encouraging the usage of a default template in which these names are explicitly imported.
For example, when you create a new project on pyscript.com, the `main.py` could be this:
```python
from pyscript import display, HTML, when

print('hello world')
```

This has two additional advantages: 
- 3rd-party tools stop complaining that `display` & co. are not defined
- it becomes obvious that the documentation of those functions live in the `pyscript` package

### Calling functions from event handlers

Currently, the following works:
```python
# main.py
def onclick():
    print('click!')
```
```html
<py-script src="main.py"></py-script>
<button py-click="onclick()">Click me</button>
```

With the new system, it becomes the following:
```html
<py-script src="main.py"></py-script>
<button py-click="main.onclick()">Click me</button>
```

**Comment by fpliger** [comment]
Thank you for this @antocuni ! This is quite helpful to set up a discussion and it's probably the right time for us to decide on this.

My initial first reacion to this is that I'm overall +1 on the proposal mostly because it makes things explicit. Honestly I'd have preferred less verbosity but I think we haven't been able to find a good compromise that doesn't make things more confusing. With that said, I probably need to sit on it a little...

**Reply by WebReflection** [reply]
> Honestly I'd have preferred less verbosity but I think we haven't been able to find a good compromise that doesn't make things more confusing.

That applies to `from polyscript import XWorker` too, as well as other fields we recently discussed :wink: 

@antocuni FWIW I never planned to have those globals in *polyscript* and even less in *pyscript* (next) ... it'll be all `from pyscript import display, XWorker` and friends and I think that's the only reasonable way forward so ... +1 here too.

**Reply by fpliger** [reply]
> That applies to from polyscript import XWorker too, as well as other fields we recently discussed 😉

Yes! I'm really hoping to write my thoughts about that later today so we can start a discussion about that too. It's also important to acknowledge the layers you hinted at above: `polyscript` and `pyscript`.

**Reply by WebReflection** [reply]
those will be two different layers with different docs too but polyscript should always be explicit ... magic happens in custom types (pyscript) and yet I fully agree magic globals aren't cool, just confusion and future issues, so I hope we can keep at least *core* (polyscript) explicit, and eventually discuss magic in pyscript, but you seem to agree explicit is always better, to which I agree.

**Comment by JeffersGlass** [comment]
Thank you for this @antocuni! In general I really like this proposal, and in that spirit, I will now try to poke holes in it 😅. Or really, there are some edge cases that I think we'll want to account for.

I will try to keep the examples to the real-world - most of them draw from things I have done or used in assorted demos and blog posts, wherever possible.

The first case is below; others I will put in separate comments to make it easier to discuss them separately, if wanted.

## Case A: Re-Using `src`

What should happen in the following scenario?

```html
<py-script src="foo.py"></py-script>
<py-script src="foo.py"></py-script>
```

In the current release, this executes the code of `foo.py` twice in the "\_\_main\_\_ namespace". In my understanding, with the new implementation, the second tag has functionally no effect.

_Why would anyone do this?_ Because we've given the location of PyScript tags semantic meaning. Because of things that depend on the location of the a tag, like `display()`, [Script Tag Plugins](https://github.com/pyscript/pyscript/blob/1b9cb562ec613ce1a5137770c330917c9ed3d147/pyscriptjs/src/plugin.ts#L60-L80) and soon [`document.currentScript`](https://github.com/pyscript/polyscript/blob/2f4da727695a15cd04682c6058e9ec3aac4ab30c/docs/README.md?plain=1#L208-L212), using the same script source in multiple locations has value.

Consider this real-life situation, which was how I originally put together my [Advent of Code in PyScript](https://jeff.glass/post/advent-of-code-2022/) page last year.

```html
<div data-code="day_1">
    <p>Day one was easy!</p>
    <py-script src="create_button_and_output.py"></py-script>
</div>
...
<div data-code="day_2">
    <p>Day two was a lot harder...</p>
    <py-script src="create_button_and_output.py"></py-script>
</div>
```

`create_run_button.py` creates a button and output div in the location of the tag, and wired them up to run some code the button was clicked (with some assistance from a Plugin, but could use `document.currentScript` in the Polyscript release). The behavior depends on the location of the script tag and its surrounding tag.

Is this the best coding practice? Probably not - I ended up re-writing that page using [Hugo's templating system](https://github.com/JeffersGlass/hugo-site/blob/a2cde716d93832cabc7cc6593db482f26fdfbbb9/layouts/shortcodes/adv2022.html#L20-L31). But still that a tag does 'nothing' just because it shares a `src` attribute with another tag on the page is a confusing error state

_So what to do? Some Options:_

  - A1: Show an error if two tags on the same page have the same `src`, or sources that resolve to the same module name (i.e. they end up in the same path in EM Filesystem)
  - A2: Don't change anything, handle with documentation
  - A3: `<py-script src="foo.py"></py-script>` becomes equivalent to the following - existing modules are 're-executed' if they are used as src twice:
  
```python
fetch_and_save("foo.py")

if 'foo' in sys.modules:
    src = open('foo.py').read()
    exec(src, sys.modules['foo'])
else:
    import foo
```

We could debate whether modules should actually be re-fetched in this case, or just re-exected from the local copy.

  - A4 Something else

**Reply by WebReflection** [reply]
I am not sure I understand the issue (with polyscript) ... you can consider two `<script type="py" src="component.py"></script>` placed in two different parts of the page and that's just fine? both will have a different `currentScript` and a different `currentScript.target` (if not specified otherwise as attribute of the script) and the only thing they share is the same *pyodide* as it's always been in pyscript.

If you don't want to share that, you can use the `env` attribute and make it contextual, but you'd be bootstrapping pyodide (or any other) twice ... so what is the issue you are trying to solve ... or better, do you have any issue with *polyscript* ?

**Reply by JeffersGlass** [reply]
No issue with Polyscript, which I think should say as is. I'm using that feature of Polyscript to highlight that we might want to change some thing about this proposal, as described above.

**Reply by antocuni** [reply]
> What should happen in the following scenario?
> 
> ```
> <py-script src="foo.py"></py-script>
> <py-script src="foo.py"></py-script>
> ```
[cut]

I think it is useful to look at what JS does; imagine two have two js files:
```js
// foo.js
console.log("foo");

// bar.js
console.log("bar");
```

```html
<script src="foo.js"></script>
<script src="foo.js"></script>

<script src="bar.js" type="module"></script>
<script src="bar.js" type="module"></script>
```

If I'm not mistaken, the code above prints `foo` twice and `bar` once: i.e., currently `<py-script>` behaves as the "old-style" `<script>` tags, but I would like to make it behaving as "new-style" `<script type="module">` tags.


> In the current release, this executes the code of `foo.py` twice in the "__main__ namespace". In my understanding, with the new implementation, the second tag has functionally no effect.

yes

> _Why would anyone do this?_ Because we've given the location of PyScript tags semantic meaning. Because of things that depend on the location of the a tag, like `display()`, [Script Tag Plugins](https://github.com/pyscript/pyscript/blob/1b9cb562ec613ce1a5137770c330917c9ed3d147/pyscriptjs/src/plugin.ts#L60-L80) and soon [`document.currentScript`](https://github.com/pyscript/polyscript/blob/2f4da727695a15cd04682c6058e9ec3aac4ab30c/docs/README.md?plain=1#L208-L212), using the same script source in multiple locations has value.
> 
> Consider this real-life situation, which was how I originally put together my [Advent of Code in PyScript](https://jeff.glass/post/advent-of-code-2022/) page last year.
> 
> ```
> <div data-code="day_1">
>     <p>Day one was easy!</p>
>     <py-script src="create_button_and_output.py"></py-script>
> </div>
> ...
> <div data-code="day_2">
>     <p>Day two was a lot harder...</p>
>     <py-script src="create_button_and_output.py"></py-script>
> </div>
> ```

yes, this kind of usage would stop working. But it is easy to recreate using functions; something like this:
```html
<py-script src="utils.py"></py-script>

<div>
    <p>Day one</p>
    <py-script>utils.create_button_and_output()</py-script>
</div>
<div>
    <p>Day two</p>
    <py-script>utils.create_button_and_output()</py-script>
</div>
```

It's also a much better development style :)


[cut]
> _So what to do? Some Options:_
> 
> * A1: Show an error if two tags on the same page have the same `src`, or sources that resolve to the same module name (i.e. they end up in the same path in EM Filesystem)
> * A2: Don't change anything, handle with documentation

I *might* be tempted to say `A1`, but I think it's just better to do the same as JS and do `A2`.

> * A3: `<py-script src="foo.py"></py-script>` becomes equivalent to the following - existing modules are 're-executed' if they are used as src twice:
> 
> ```python
> fetch_and_save("foo.py")
> 
> if 'foo' in sys.modules:
>     src = open('foo.py').read()
>     exec(src, sys.modules['foo'])
> else:
>     import foo
> ```

ok, now I will have nightmares 😂

**Reply by WebReflection** [reply]
>  but I would like to make it behaving as "new-style" <script type="module"> tags.

I respectfully disagree here ... the source is not strictly a module, rather a behavior like htc files used to be.

If we go your road, and if I understood it correctly, it will be impossible to define pyscript or mpyscript components because only the first one would work and others do nothing.

```html
<ul>
  <!-- these are two PyScript components, same way Custom Elements would be -->
  <li><script type="py" src="li_component.py"></script></li>
  <li><script type="py" src="li_component.py"></script></li>
</ul>
```
Importing modules is a Python affair, here we're importing and executing files just like `<script src="..">` does and that's **desired** because `type="module"` is a completely different beast and has different purposes so I would be always -1 there as ~~it dooms the project for no use-case or real-world benefit to me ... quite the opposite.~~ (more thoughts added after)

**edit** more on this ... if you expect a script to run only once you write that scirpt with that *src* **only once** ... having the ability or guards against multiple scripts with the same source is bad under all circumstances ... if a user did that it means it expect that to work, not to behave like a *JavaScript module*, imho.

**Reply by WebReflection** [reply]
more on the topic ... in JS modules do execute once but their exports are always usable. If the intent here is to cache executions of the same source per interpreter (env) then it's OK but the issue here is `document.currentScript` that cannot be cached.

Specially because of `document.currentScript` we really **want** scripts to execute every time, even if their methods will override previous methods **but**, admittedly, there could be at that point race conditions in references so we need to carefully decide how to proceed there because if somebody does this:

```py
from js import document

currentScript = document.currentScript

method():
  currentScirpt.target.textContent = 123
```

then the last script with the same source will indeed override previous script with `method` and the `currentScript` reference will be lost ... I might create a demo/test-case to investigate this as it is potentially a mess to deal with, although I don't see any other standard way to retrieve the current executing script *but* there could be a way to attach to it methods previously exported for one and not for others as after all methods is what we're after.

**Reply by WebReflection** [reply]
Also on the topic: the `env` can disambiguate by bootstrapping a new python-env per each component ... this easily doesn't scale with dozen components (or thousand) on the page so we might try to see if a hot-swap of previous methods per each py/script is possible ... as in:
  * the script is found
  * the source file runs
  * globals in that run are extracted and weakly attached to the script
  * when the script execute a callback, it uses those globals
  * next script will do the same so previous script is safe to execute anything it likes

I think this might be a great solution, if possible, but I am afraid I am not sure I fully understand how retrieving globals works and/or if it's possible to hot-swap these on the fly on each *click*

**Reply by antocuni** [reply]
> If we go your road, and if I understood it correctly, it will be impossible to define pyscript or mpyscript components because only the first one would work and others do nothing.
> 
> ```
> <ul>
>   
>   <li><script type="py" src="li_component.py"></script></li>
>   <li><script type="py" src="li_component.py"></script></li>
> </ul>
> ```

I lost you here.
What is a "pyscript component"? How do you envision the content of `li_component.py` in the example above?

**Reply by WebReflection** [reply]
the `li_component.py` could be something like:

```python
from js import document

target = document.currentScript.target

# actually do something meaningful with the target
target.innerHTML = "<button>click me</button>"

def click_handler(event):
    print(event.type)
    event.currentTarget.parentNode.dataset.clicked = True

target.firstElementChild.addEventListener("click", click_handler)
```

silly example, but this would work already now.

so each list component can bootstrap itself, attach listeners to itself, react, and so on ... with your idea, none of this would be possible

**Reply by WebReflection** [reply]
P.S. please note I've used `event.currentTarget.parentNode` instead of `target` because of the issue we're discussing when same code runs multiple times ... it's the *when* it runs that matters here for `currentScript` reference sake and I can't think of a way to avoid re-running code when *components* are desirable

**Reply by antocuni** [reply]
Ok, so it's basically the same example as Jeff's `create_button_and_run.py`.

Personally, I don't think that this is a code pattern that we should encourage: here we are basically using the file as the "unit of reusable code", but this is something that it's very surprising from the perspective of a python programmer.

The semantics that you are proposing are very similar to C's `#include` and php's `include`, but it's a concept that it doesn't really exist in the Python world.
In Python, there is a 1:1 mapping between "one .py file" and "one module", and all the code which is around expects that "module level statements" are executed only once, at initialization time.

The most pythonic way to achieve what you propose is to wrap the code into a function, and call the function multiple times.

**Reply by WebReflection** [reply]
we said we want to be closer to Web standards and there's no other way to meaningfully use `currentScript` ... that was the key to unlock otherwise impossible to have "classic pyscript" scenarios because the script can be anywhere and polyscript core allows that and it should, as that's how scripts work.

In short, unless you find a solution for `document.currentScript` I don't buy your argument of making it more pythonic by behaving like a JavaScript module and here the script is declarative on HTML which has already nothing to do with Python or the pythonic way, it's just how HTML and the Web works.

If we kill currentScript we can also already stop working on polyscript and pyscript next, imho.

**Reply by WebReflection** [reply]
On the other hand, we can reduce all features as much as we like with PyScript custom type, implementing all defensive and not desired logic we think should never exist but current *polyscript*, and Jeff seems to agree with me, works just fine as it is and it shouldn't change its features ... with `pyscript` module we can decide also which `currentScript` is running or put a `target` there directly without affecting *polyscript* so that other people, from **other progrmaming languages**, can use *polyscript* as they please, not as Python and people using Python want them to use the project (Lua, Ruby, R, others ...).

That is: *polyscript* has nothing to do with the pythonic way and we should discuss only things we want in *pyscript*, as *polyscript* custom type, not how core should work. All arguments about *pythonic* imho should be ignored for *polyscript*, but of course if it needs to provide some utility to make *pyscript* more pythonic, I'd be fully OK with adding those features, if these don't break, or limit, existing features.

I hope my position is reasonable and clear.

**Reply by antocuni** [reply]
> That is: polyscript has nothing to do with the pythonic way and we should discuss only things we want in pyscript,

that's exactly what I'm discussing: pyscript.

**Reply by WebReflection** [reply]
My bad then ... I am OK with us reducing features and potentials in the name of the *pythonic* way or *least surprise* behavior but I will need to add flags in *polyscript* to disallow mmultiple file evaluation or cache them as module or implement automatic import of stuff that's automatically saved in the root folder so ... once we decide what we want PyScript Next to be, let's write down what's required from *polyscript* to allow PyScript to work as we want/like/prefer.

Until that I am a bit stuck with progress around it though so I hope we'll find the way we want it to be sooner than later.

**Comment by JeffersGlass** [comment]
## Case B: `if __name__ == '__main__'`

There have been a number of users on the `[pyscript]` tag on Stack Overflow confused about using  `if __name__ == '__main__'` in PyScript [[1]](https://stackoverflow.com/questions/72620014/pyscript-mqtt-publishing-message) [[2]](https://stackoverflow.com/questions/74533570/pyscript-modulenotfounderror-no-module-named-cv2) [[3]](https://stackoverflow.com/questions/73351630/jsexceptionpythonerror-jsexception-error-object-has-already-been-destroyed-t) [[4]](https://stackoverflow.com/questions/72620014/pyscript-mqtt-publishing-message). Possibly from copying code over from 'normal' Python straight into a PyScript tag. Of course in the current release, this is always False.

With the proposed change here, this idiom be True in an an inline tag, but False in a tag with a `src`, correct?

I think the place where this would bite users is that when they start writing code in an inline tag and then move it to an external source, the behavior of their code would possibly change. Or when a user posts a question and just their Python code the answer to their question might depend on whether the tag is inline or has an external source.

_So what to do? Some Options:_
  - B1: Do nothing, handle with documentation.
  - B2: Show a warning if we detect the user is using something that looks like "__main__" in an externally sourced file. This is appealing to me but troubling in implementation - we could do something simple like looking for `__main__` in the src or look for a string literal `__main__` in the AST or... I'm not sure. Might be prone to false positives.
  - B3: Something else.

**Reply by antocuni** [reply]
> There have been a number of users on the `[pyscript]` tag on Stack Overflow confused about using `if __name__ == '__main__'` in PyScript [[1]](https://stackoverflow.com/questions/72620014/pyscript-mqtt-publishing-message) [[2]](https://stackoverflow.com/questions/74533570/pyscript-modulenotfounderror-no-module-named-cv2) [[3]](https://stackoverflow.com/questions/73351630/jsexceptionpythonerror-jsexception-error-object-has-already-been-destroyed-t) [[4]](https://stackoverflow.com/questions/72620014/pyscript-mqtt-publishing-message). Possibly from copying code over from 'normal' Python straight into a PyScript tag. Of course in the current release, this is always False.

uh? No, this is not true. In the current release, `__name__` is *always* `'__main__'`: [proof](https://antocuni.pyscriptapps.com/if-name-main/latest/).

It seems that in the SO questions that you linked above, the problems were elsewhere, but not in the `if __name__` part.

> With the proposed change here, this idiom be True in an an inline tag, but False in a tag with a `src`, correct?

correct

> I think the place where this would bite users is that when they start writing code in an inline tag and then move it to an external source, the behavior of their code would possibly change. Or when a user posts a question and just their Python code the answer to their question might depend on whether the tag is inline or has an external source.

I admit that this is not nice, but at the same time it seems the lesser evil, compared to the current situation.
Maybe we could improve by not using the name `__main__` at all: the code contained inside `index.html` get `__name__ == 'index'`, and other modules can import its content by doing `import index`.
This might actually be a good idea and be *less* surprising than having to `import __main__`, which is a bit obscure.

The biggest cons is that if you have a file named `index.py`, it would basically be ignored: as a possible workaround, pyscript.com could warn the users if they try to create a file with such a name.

**Comment by WebReflection** [comment]
> With the new system, it becomes the following:

```html
<py-script src="main.py"></py-script>
<button py-click="main.onclick()">Click me</button>
```

this one I don't understand ... but we're at the fourth iteration around events in polyscript, and I think we're good the way it is now ... the `env` once again disambiguate whenever that's desirable so you can use `env="main"` and `py-env="main"` to grab exported functions within the `py-click`.

If you're talking about pyscript classic then I think it's a bit late to change stuff in a breaking way there, but that's also not my call.

**Reply by WebReflection** [reply]
P.S. I think this is worth throwing in here ...

![Screenshot from 2023-07-27 18-08-58](https://github.com/pyscript/pyscript/assets/85749/74ed14f9-a439-4afa-ae43-b191b019ad31)

I've been contacted via DM on "Xwitter" as that was a desired feature ... I've explained our position around rolling back that feature (the usable `event` and the multiple args via invoke/evaluation) and we found a way through `dataset` (hence `data-xxx`) attributes to pass somehow data around to the listener.

I still believe we should run polls and ask our users what they think/prefer as feature because it's not the first time we *assume* what they want but I haven't seen any real/concrete user research around our assumptions.

I like the Jeff approach to take at least SO issues as metric, it's still something better than "*our gut feeling*" (or just personal opinion) but I think we should play better at asking feedback via social media or by any other mean that could reach as many users as possible.

**Reply by antocuni** [reply]
> ```
> <py-script src="main.py"></py-script>
> <button py-click="main.onclick()">Click me</button>
> ```
>   
> this one I don't understand ... but we're at the fourth iteration around events in polyscript, and I think we're good the way it is now ... 

in this example I put the parenthesis because I was comparing with pyscript-classic, but of course in pyscript-next they won't be there.

As for the name, I always advocated that `py-click` should allow a dotted expression :)

**Reply by WebReflection** [reply]
dotted expressions are already allowed in polyscript ... it's that `main` that I've no idea where it comes from ... is it derived from the file name?

**Reply by antocuni** [reply]
In my proposal:
```
<py-script src="main.py"></py-script>
```

would be equivalent to:
```
<py-script>
    fetch_and_save('main.py')
    import main
    # now main is a valid name of the "global" scope inside index.html
</py-script>
```

So yes, it's derived from the filename.

**Reply by WebReflection** [reply]
but in events, that import would be ~~explicit~~ **implicit** ... so what happens when another script does:

```html
<py-script>
def main():
  print("busted?")
</py-script>

<!-- or -->

<py-script>
main = SomeClass()
</py-script>
```

???

**Reply by antocuni** [reply]
yes, in that case you have a name clash. It's exactly the same as this:
```
<py-script>
import main
# here main points to a module

def main():
    ...
# here main points to the function defined above

main = SomeClass()
# here main points to the instance created above
</py-script>
```

**Reply by WebReflection** [reply]
so what are we solving here? :thinking:

**Reply by WebReflection** [reply]
Also, to clarify, other interpreters don't have the ability to save files and use these as imports ... I might have said this before, but we should stop thinking about Python only way of doing things and any time we propose something think if that's even possible elsewhere.

Here I see changes to have an `import root` before running the `root.thing` event and that might not be possible with other interpreters, or it might be async (webR) or not possible at all (Ruby-wasm-wasi or wasmoon).

**Reply by WebReflection** [reply]
P.S. to clarify all scenarios, we might want to provide a different `runEvent` in polyscript for Python cases, but that would branch documentation and use cases and behavior across other interpreters and I am not sure that's a good idea.

If we make Python the only first-class citizen of Polyscript then we should not have other interpreters in the roadmap as these will be less interesting by design, just my 2 cents on the matter.

**Reply by WebReflection** [reply]
OK, as clarified [in here](https://github.com/pyscript/pyscript/discussions/1612#discussioncomment-6575061), it looks like I should stop mentioning *polyscript* in this discussion as we're just deciding how PyScript Next should be.

I'd be OK in having it be the way we want/like/prefer as PyScript Next will need its own documentation anyway so that branching there is not an issue.

**Reply by antocuni** [reply]
> If we make Python the only first-class citizen of Polyscript then we should not have other interpreters in the roadmap as these will be less interesting by design, just my 2 cents on the matter.

this is a good point and it's worth its own (big) discussion. Personally, my position is that every new language that we want to support introduces complications and augments the number of trade-offs that we have to do (and my gut feeling is that the number of trade-offs scales non-linearly with the number of languages, but I can't prove that 😅).

So, we should sit down and decide whether we want that all languages are equal, or maybe that [all languages are equal, but python is more equal than others](https://www.dictionary.com/browse/all-animals-are-equal--but-some-animals-are-more-equal-than-others). 

I have my own personal preference of course, but actually both options are reasonable: the important thing is to understand what are the pros and cons of each one and make an explicit and conscious choice, which I think we never did.

**Reply by WebReflection** [reply]
I did make a conscious choice behind *polyscript* and it can already work in every PL I know [thanks to its primitives](https://pyscript.github.io/polyscript/#interpreter-features):

  * `run`
  * `runAsync`
  * `runEvent`

That's literally all it enables and if every PL starts messing around with just these primitives there's not much we can do.

There is only one particular special primitive that *should* be possible to enable in all PLs but we're currently falling back to globals:

  * `registerJSModule`

That's already a different thing in Python interpreters VS others, and I hope it'll be the only one.

The extra `writeFile` is also for fetch but the way it gets read or imported is PL dependent, so that's good too, whenever it's supported.

That's why I've said *polyscript* should be kept language agnostic, and there's nothing really different in those primitives among languages, with the exception that *webR* right now can't execute code *synchronously*, only *asynchronously*, but that's thier choice, so that using `run` or `runAsync` might either do the same or throw when `run` is used, like we throw in Ruby which has the *experimental* flag.

Everything else should be kept confined in the custom type world where any PL or user or team can use these primitives and the little orchestration behind workers and DOM features without dictating how every other PL should look or work like.
